File: sth/sales_sth/report/laporan_penerimaan_tbs/laporan_penerimaan_tbs.js
// Copyright (c) 2024, Your Company and contributors
// For license information, please see license.txt

frappe.query_reports["Laporan Penerimaan TBS"] = {
    "filters": [
        {
            "fieldname": "laporan",
            "label": __("Laporan"),
            "fieldtype": "Select",
            "options": ["Laporan Penerimaan TBS", "Rekap Penerimaan TBS"],
            "default": "Laporan Penerimaan TBS",
            "reqd": 1,
            "on_change": function() {
                let report = frappe.query_report;
                let laporan_type = report.get_filter_value('laporan');
                
                // Show/hide filters based on report type
                if (laporan_type === "Rekap Penerimaan TBS") {
                    report.get_filter('tbs').toggle(false);
                    report.get_filter('supplier').toggle(false);
                    report.get_filter('tanggal_dari').toggle(false);
                    report.get_filter('tanggal_sampai').toggle(false);
                    report.get_filter('company').toggle(true);
                    report.get_filter('tanggal').toggle(true);
                } else {
                    report.get_filter('tbs').toggle(true);
                    report.get_filter('supplier').toggle(true);
                    report.get_filter('tanggal_dari').toggle(true);
                    report.get_filter('tanggal_sampai').toggle(true);
                    report.get_filter('company').toggle(false);
                    report.get_filter('tanggal').toggle(false);
                }
                
                report.refresh();
            }
        },
        {
            "fieldname": "tbs",
            "label": __("TBS"),
            "fieldtype": "Select",
            "options": ["", "External", "Internal"],
            "default": "",
            "reqd": 0
        },
        {
            "fieldname": "supplier",
            "label": __("Supplier"),
            "fieldtype": "Link",
            "options": "Supplier",
            "reqd": 0,
            "get_query": function() {
                return {
                    "filters": {
                        "disabled": 0
                    }
                }
            }
        },
        {
            "fieldname": "tanggal_dari",
            "label": __("Tanggal Dari"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_start(),
            "reqd": 0
        },
        {
            "fieldname": "tanggal_sampai",
            "label": __("Tanggal Sampai"),
            "fieldtype": "Date",
            "default": frappe.datetime.month_end(),
            "reqd": 0
        },
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 0,
            "hidden": 1
        },
        {
            "fieldname": "tanggal",
            "label": __("Tanggal"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 0,
            "hidden": 1
        }
    ],
    
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        
        let laporan_type = frappe.query_report.get_filter_value('laporan');
        
        if (laporan_type === "Rekap Penerimaan TBS") {
            // Format for Rekap report
            if (column.fieldname !== "driver_name") {
                if (value && !isNaN(parseFloat(value))) {
                    value = `<div style="text-align: right">${parseFloat(value).toLocaleString('id-ID', {minimumFractionDigits: 2, maximumFractionDigits: 2})}</div>`;
                }
            }
            
            // Bold formatting for total rows
            if (data && (data.driver_name === "Total Internal" || data.driver_name === "Total External")) {
                value = `<div style="font-weight: bold">${value}</div>`;
            }
        } else {
            // Format for Laporan Penerimaan TBS
            if (column.fieldname == "bruto" || 
                column.fieldname == "tara" || 
                column.fieldname == "netto" ||
                column.fieldname == "potongan" ||
                column.fieldname == "berat_normal") {
                if (value && !isNaN(parseFloat(value))) {
                    value = `<div style="text-align: right">${parseFloat(value).toFixed(2)}</div>`;
                }
            }
        }
        
        return value;
    },
    
    "onload": function(report) {
        // Initialize filter visibility
        let laporan_type = report.get_filter_value('laporan');
        
        if (laporan_type === "Rekap Penerimaan TBS") {
            report.get_filter('tbs').toggle(false);
            report.get_filter('supplier').toggle(false);
            report.get_filter('tanggal_dari').toggle(false);
            report.get_filter('tanggal_sampai').toggle(false);
            report.get_filter('company').toggle(true);
            report.get_filter('tanggal').toggle(true);
        } else {
            report.get_filter('company').toggle(false);
            report.get_filter('tanggal').toggle(false);
        }
    }
};

File: sth/sales_sth/report/laporan_penerimaan_tbs/laporan_penerimaan_tbs.py
# Copyright (c) 2024, Your Company and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from datetime import datetime
import calendar

def execute(filters=None):
	laporan_type = filters.get("laporan", "Laporan Penerimaan TBS")
	
	if laporan_type == "Rekap Penerimaan TBS":
		columns = get_rekap_columns(filters)
		data = get_rekap_data(filters)
	else:
		columns = get_columns()
		data = get_data(filters)
	
	return columns, data

def get_columns():
	"""Define kolom-kolom untuk laporan penerimaan TBS"""
	return [
		{
			"fieldname": "posting_date",
			"label": _("Tanggal"),
			"fieldtype": "Date",
			"width": 100
		},
		{
			"fieldname": "weight_in_time",
			"label": _("Jam Masuk"),
			"fieldtype": "Time",
			"width": 100
		},
		{
			"fieldname": "weight_out_time",
			"label": _("Jam Keluar"),
			"fieldtype": "Time",
			"width": 100
		},
		{
			"fieldname": "supplier",
			"label": _("Nama Supplier"),
			"fieldtype": "Link",
			"options": "Supplier",
			"width": 150
		},
		{
			"fieldname": "ticket_number",
			"label": _("Tiket No"),
			"fieldtype": "Data",
			"width": 120
		},
		{
			"fieldname": "license_number",
			"label": _("Kode Alat"),
			"fieldtype": "Data",
			"width": 100
		},
		{
			"fieldname": "bruto",
			"label": _("Berat Masuk"),
			"fieldtype": "Float",
			"width": 120
		},
		{
			"fieldname": "tara",
			"label": _("Berat Keluar"),
			"fieldtype": "Float",
			"width": 120
		},
		{
			"fieldname": "netto",
			"label": _("Berat Bersih"),
			"fieldtype": "Int",
			"width": 120
		},
		{
			"fieldname": "potongan",
			"label": _("Potongan (kg)"),
			"fieldtype": "Int",
			"width": 120
		},
		{
			"fieldname": "berat_normal",
			"label": _("Berat Normal"),
			"fieldtype": "Int",
			"width": 120
		},
		{
			"fieldname": "driver_name",
			"label": _("Nama Sopir"),
			"fieldtype": "Data",
			"width": 150
		}
	]

def get_data(filters):
	"""Ambil data dari doctype Timbangan berdasarkan filter"""
	
	conditions = get_conditions(filters)
	
	data = frappe.db.sql("""
		SELECT
			posting_date,
			weight_in_time,
			weight_out_time,
			supplier,
			ticket_number,
			license_number,
			bruto,
			tara,
			netto,
			potongan_sortasi * netto / 100 as potongan,
			netto - (potongan_sortasi * netto / 100) as berat_normal,
			driver_name
		FROM
			`tabTimbangan`
		WHERE
			receive_type IN ('TBS Internal', 'TBS Eksternal')
			AND docstatus = 1
			{conditions}
		ORDER BY
			posting_date, weight_in_time
	""".format(conditions=conditions), filters, as_dict=1)
	
	return data

def get_conditions(filters):
	"""Build kondisi WHERE berdasarkan filter"""
	conditions = []
	
	if filters.get("tbs"):
		if filters.get("tbs") == "External":
			conditions.append("AND receive_type = 'TBS Eksternal'")
		elif filters.get("tbs") == "Internal":
			conditions.append("AND receive_type = 'TBS Internal'")
	
	if filters.get("supplier"):
		conditions.append("AND supplier = %(supplier)s")
	
	if filters.get("tanggal_dari"):
		conditions.append("AND posting_date >= %(tanggal_dari)s")
	
	if filters.get("tanggal_sampai"):
		conditions.append("AND posting_date <= %(tanggal_sampai)s")

	if filters.get("company"):
		conditions.append("AND company = %(company)s")

	if filters.get("unit"):
		conditions.append("AND unit = %(unit)s")
	
	return " ".join(conditions)

def get_rekap_columns(filters):
	"""Generate dynamic columns for Rekap report based on the selected month"""
	tanggal = filters.get("tanggal")
	
	if not tanggal:
		tanggal = datetime.today().date()
	elif isinstance(tanggal, str):
		tanggal = datetime.strptime(tanggal, "%Y-%m-%d").date()
	
	year = tanggal.year
	month = tanggal.month
	
	# Get number of days in the month
	days_in_month = calendar.monthrange(year, month)[1]
	
	columns = [
		{
			"fieldname": "driver_name",
			"label": _("Tanggal"),
			"fieldtype": "Data",
			"width": 150
		}
	]
	
	# Add columns for each day
	for day in range(1, days_in_month + 1):
		columns.append({
			"fieldname": f"day_{day}",
			"label": str(day),
			"fieldtype": "Float",
			"width": 100,
			"align": "left"
		})
	
	# Add Total column
	columns.append({
		"fieldname": "total",
		"label": _("Total"),
		"fieldtype": "Float",
		"width": 120
	})
	
	return columns

def get_rekap_data(filters):
	"""Get recap data grouped by driver and date"""
	tanggal = filters.get("tanggal")
	
	if not tanggal:
		tanggal = datetime.today().date()
	elif isinstance(tanggal, str):
		tanggal = datetime.strptime(tanggal, "%Y-%m-%d").date()
	
	year = tanggal.year
	month = tanggal.month
	
	# Get first and last day of the month
	first_day = datetime(year, month, 1).date()
	last_day = datetime(year, month, calendar.monthrange(year, month)[1]).date()
	
	# Query data for Internal TBS
	internal_data = frappe.db.sql("""
		SELECT
			driver_name,
			DAY(posting_date) as day,
			SUM(netto) as total_netto
		FROM
			`tabTimbangan`
		WHERE
			receive_type = 'TBS Internal'
			AND docstatus = 1
			AND posting_date BETWEEN %s AND %s
		GROUP BY
			driver_name, DAY(posting_date)
		ORDER BY
			driver_name, DAY(posting_date)
	""", (first_day, last_day), as_dict=1)
	
	# Query data for External TBS
	external_data = frappe.db.sql("""
		SELECT
			driver_name,
			DAY(posting_date) as day,
			SUM(netto) as total_netto
		FROM
			`tabTimbangan`
		WHERE
			receive_type = 'TBS Eksternal'
			AND docstatus = 1 
			AND posting_date BETWEEN %s AND %s
		GROUP BY
			driver_name, DAY(posting_date)
		ORDER BY
			driver_name, DAY(posting_date)
	""", (first_day, last_day), as_dict=1)
	
	# Process data
	result = []
	
	# Process Internal data
	internal_drivers = process_driver_data(internal_data)
	result.extend(internal_drivers)
	
	# Add Total Internal row
	if internal_drivers:
		total_internal = calculate_total_row(internal_drivers, "Total Internal")
		result.append(total_internal)
	
	# Process External data
	external_drivers = process_driver_data(external_data)
	result.extend(external_drivers)
	
	# Add Total External row
	if external_drivers:
		total_external = calculate_total_row(external_drivers, "Total External")
		result.append(total_external)
	
	return result

def process_driver_data(data):
	"""Process raw data into driver-based rows"""
	drivers = {}
	
	for row in data:
		driver = row.get("driver_name") or "Unknown"
		day = row.get("day")
		netto = row.get("total_netto", 0)
		
		if driver not in drivers:
			drivers[driver] = {"driver_name": driver, "total": 0}
		
		drivers[driver][f"day_{day}"] = netto
		drivers[driver]["total"] += netto
	
	return list(drivers.values())

def calculate_total_row(driver_rows, label):
	"""Calculate total row from driver rows"""
	total_row = {"driver_name": label, "total": 0}
	
	for driver in driver_rows:
		for key, value in driver.items():
			if key.startswith("day_"):
				if key not in total_row:
					total_row[key] = 0
				total_row[key] += value
			elif key == "total":
				total_row["total"] += value
	
	return total_row